### DISCUSSION TITLE
How to Extend theme in @rneui/themed (v4)

### DISCUSSION THREAD

**Discussion by lucksp** [react-native-elements/react-native-elements#3441] [Q&A]
I do not want to use the [default `Colors` theme shape type](https://reactnativeelements.com/docs/customization/color) because it does not fit our needs.   I added a `react-native-elements.d.ts` file to extend the theme, as suggested in docs, but it [brings up the docs](https://reactnativeelements.com/docs/3.4.2/customization#typescript-definitions-extending-the-default-theme) to v3.4.

// react-native-elements.d.ts
```
import '@rneui/themed';
import { ColorShape } from '../lib/theme/colors/theme.type';

declare module '@rneui/themed' {
    export interface CreateThemeOptions {
        myColors: ColorShape;
    }
}
```

and then in my App, i add the ThemeProvider, and inject the new theme as prop:

// app.tsx
```
import { ThemeProvider, FullTheme, createTheme } from '@rneui/themed';

const withThemeProvider = (Story, context) => {
  const colors = useDarkMode() ? dark : light;

  const themeObj = createTheme({
    myColors: colors
  })

  return (
    <ThemeProvider theme={themeObj}>
        {children}
    </ThemeProvider>
  );
};
```

What I am seeing is that the `ThemeProvider` in V4 requests theme to be of type: `theme?: CreateThemeOptions`, where in v3, the theme is of type `(property) theme?: {} | undefined`.

So v4 is much more strict about this.  However, I still thought that by creating a new `interface CreateThemeOptions` above, that we could add a new type in.  Typescript complains:

> Argument of type '{ myColors: ColorShape; }' is not assignable to parameter of type 'CreateThemeOptions'.
  Object literal may only specify known properties, and 'myColors' does not exist in type 'CreateThemeOptions'.ts(2345)

How can I add this custom theme shape?

**Comment by arpitBhalla** [comment]
How about just extend the `Colors` or `FullTheme` with myColors? actually in v4 you can specify `darkColors` & `lightColors` separately, and update it using `updateTheme`

Also can you share `ColorShape` type?

**Reply by lucksp** [reply]
I am not sure the extending as the documents imply on v4 works...

The `CreateThemeOptions` which have the `lightColors` and `darkColors` are [so closely tied](https://github.com/react-native-elements/react-native-elements/blob/next/packages/themed/src/config/ThemeProvider.tsx#L22-L25) to the RNE [`Colors` shape](https://github.com/react-native-elements/react-native-elements/blob/next/packages/base/src/helpers/colors.ts) that it doesn't seem like an ideal option.

My custom `ColorShape` is:
```
export interface ColorShape {
  backgroundColor: string;
  constants: {
    black: string;
    bronze: string;
    darkBlue: string;
    fluidBlue: string;
    gold: string;
    orange: string;
    silver: string;
    white: string;
  };
  fluidBlue_ada: string;
  green: string;
  keylines: {
    keyline_default: string;
    keyline_light: string;
  };
  red: string;
  text: {
    ada: string;
    primary: string;
    secondary: string;
  };
  transparent: string;
  ui_color_1: string;
}
```

It's as if the `react-native-elements.d.ts` declaration isn't behaving as intended...where in v3 it works because the `theme` is just an `{}`

**Reply by arpitBhalla** [reply]
```tsx

declare module "@rneui/themed" {
  export interface Colors extends ColorShape {}
  export interface FullTheme {}
}

const theme = createTheme({
  darkColors: {
    red: "#ff0000",
    constants: {
      black: "black",
    },
  },
  lightColors: {
    red: "#00ff00",
  },
});
```

OR

```tsx

declare module "@rneui/themed" {
  export interface FullTheme {
    myColors: Partial<ColorShape>;
  }
}

const theme = createTheme({
  myColors: { red: "red", fluidBlue_ada: "demo" },
});
```

**Reply by arpitBhalla** [reply]
I don't think it should be in `react-native-elements.d.ts`, need to update docs

**Reply by lucksp** [reply]
> I don't think it should be in `react-native-elements.d.ts`, need to update docs

where should it live?

**Reply by arpitBhalla** [reply]
I am not sure, if you can try create a dir in root `@rneui` and another dir in it named `themed` and a file in it `index.d.ts` 
Or just add it where you are using createTheme (In the same file)

**Reply by arpitBhalla** [reply]
![image](https://user-images.githubusercontent.com/55053424/161838354-e8c3e02d-0f24-4063-9850-93d70f85991d.png)
This worked for me

**Reply by lucksp** [reply]
OK GREAT.  It is working in the same file, as when I try to `createTheme`, but not outside of it in a `.d.ts` file.  For now this will be fine.  

THANK YOU 🥇 


EDIT - let me try your d.ts setup

**Reply by arpitBhalla** [reply]
If you don't mind, can you update the docs?

**Reply by lucksp** [reply]
For some reason that d.ts file just wont get picked up...tried updating the tsconfig, etc...

But i can update the docs with something closer to this setup...

**Reply by arpitBhalla** [reply]
Can you share d.ts file?

**Comment by lucksp** [comment]
> Can you share d.ts file?

Sure:

<img width="888" alt="image" src="https://user-images.githubusercontent.com/15162169/161846847-74f486a0-fe27-455d-aa19-33cd98e526c5.png">


and consuming d.ts file outside of the file being used:
<img width="1074" alt="image" src="https://user-images.githubusercontent.com/15162169/161847012-029d20e3-2b1e-4a8a-aea2-5b7c28a8a6dd.png">

**Reply by arpitBhalla** [reply]
in d.ts use `import "@rneui/themed"`, the @rneui folder should contain `themed`, but in your case it's `theme`

**Comment by lucksp** [comment]
We're so close!

Here is a repo: https://github.com/lucksp/expo-storybook

What would I use in the component to access `theme.myColors` obj?  `useTheme` doesn't seem to do the trick:
- because `theme` prop in the ThemeProvider is optional for some reason, so in order to use the theme it needs optional chaining. 
- because the custom color shape is not fully recognized (see photo)
  - <img width="636" alt="image" src="https://user-images.githubusercontent.com/15162169/162001842-fcf63d81-3c76-4b06-952f-5be5c93e42dc.png">


This could relate to the setup:
- if I have ONLY `@renui/themed.d.ts` file, then the object to initialize the ThemeProvider `theme` fails as the shape is not valid
- if I `declare module '@rneui/themed'` INSIDE the same file that initializes the ThemeProvider, then the `useTheme` fails inside my consuming component because the new shape is not found.
  - if I remove anything from the same file that uses the ThemeProvider, then it doesn't recognize the `d.ts` file setting
- if I have BOTH a `themed.d.ts` file AND `declare module '@rneui/themed'` inside the theme provider file, then it works...but it feels wrong to have both.


// themed.d.ts.
```js
import "@rneui/themed";
import { ColorShape } from "../theme/colors/theme.type";

declare module "@rneui/themed" {
    export interface FullTheme {
        fluidColors: ColorShape;
    }
    // export interface CreateThemeOptions {
    //     fluidColors: ColorShape;
    // }
}
```

//then in my "preview" file:

```js
declare module '@rneui/themed' {
  export interface FullTheme {
    fluidColors: ColorShape;
  }
  // export interface CreateThemeOptions {
  //   fluidColors: ColorShape;
  // }
}

// ...and then
const withThemeProvider = (Story, context) => {
  const colors = useDarkMode() ? dark : light;

  const themeObj = createTheme({
    fluidColors: colors
  })
//  themeObj.myColors.backgroundColor is not found, `fluidColors` doesn't have any of the custom shape??

  return (
    <ThemeProvider theme={themeObj}>
        {children}
    </ThemeProvider>
  );
};
```

Now, I create a `Button` which would consume `theme.myColors`:
```js
import { Button as RNEButton, ButtonProps } from '@rneui/base';
import { ThemeContext, useTheme } from '@rneui/themed';
import { useContext } from 'react';

const Button = (props: ButtonProps) => {
  const { theme } = useTheme();
// theme.fluidColors does not exist
  const themeContext = useContext(ThemeContext);
// themeContext.theme does not contain the colors either

  return <RNEButton {...props} buttonStyle={{
    backgroundColor: '#19AFEB', // should be theme.fluidColors.backgroundColor
    borderRadius: 32,
  }}
  titleStyle={{
    color: '#FFF', // should be theme.fluidColors.text.primary
  }}
  />;
};
```

Any other ideas?  Are there other hooks to use?  Does the repo link above help?

**Reply by arpitBhalla** [reply]
Hey, didn't get a chance to go through the repo yet, will help you with this, Can you DM me on Slack?

**Reply by lucksp** [reply]
sure!  what's the best way to find you on Slack?

I can start a google hangout other wise :)

**Reply by arpitBhalla** [reply]
Your point is valid, extending theme needs to be fixed

**Reply by arpitBhalla** [reply]
Created #3444

**Reply by lucksp** [reply]
thanks for the reply.  Maybe I can help on this.

**Reply by arpitBhalla** [reply]
yes yes, you can

**Reply by arpitBhalla** [reply]
Not sure, but inside `.storybook` TS isn't recognising the extended theme, works fine anywhere.

**Reply by lucksp** [reply]
thank you for the update.

**Reply by arpitBhalla** [reply]
Hey, I have fixed that issue, can you try extending `Theme` (not `FullTheme`)

**Reply by lucksp** [reply]
thanks Arpit.  I was looking for the [release](https://github.com/react-native-elements/react-native-elements/releases) which includes PR https://github.com/react-native-elements/react-native-elements/pull/3450 but do not see it yet. What version can we install to test this out?

**Reply by arpitBhalla** [reply]
Hey, Can you try it with [Bleeding Edge](https://reactnativeelements.com/docs/installation#bleeding-edge) version

**Reply by lucksp** [reply]
hi @arpitBhalla so far bleeding edge has worked well for the ThemeProvider implementation and the `makeStyles` since it references the `FullTheme`.  

We are running into 2 things:

1. the `useTheme` hook.  

```ts
interface useThemeI {
    replaceTheme: ReplaceTheme;
    updateTheme: UpdateTheme;
    theme: {
        colors: Colors;
        mode: ThemeMode;
    } & Theme;
}
```

this is because the `theme` field doesn't reference the `FullTheme` which is created at the start in `themed.d.ts`.

What do you suggest as a possible work around?   I have tried to update my `themed.d.ts` file to:

```ts
export interface useThemeI {
    replaceTheme: ReplaceTheme;
    updateTheme: UpdateTheme;
    theme: {
      colors: Colors;
      mode: ThemeMode;
      fluidColors: ColorShape; // added this
    } & Theme;
  }
```
and
```ts
export interface useThemeI {
    replaceTheme: ReplaceTheme;
    updateTheme: UpdateTheme;
    theme: {
      colors: Colors;
      mode: ThemeMode;
    } & FullTheme;  // updated to this
  }
```
but without success...I appreciate any suggestions/thoughts.




2. the `makeStyles` hook.  While the TS seems ok, it appears our custom theme is not available when the code runs
![image](https://user-images.githubusercontent.com/15162169/165364964-baff8903-68a5-494d-9264-c2b316c4d6f8.png)

I believe it's related to the `makeStyles()` function where it does not pass in custom theme fields.  Possibly it should look like:
```ts
export const makeStyles = (styles) => (props = {}) => {
    const { theme } = useTheme();
    return useMemo(() => {
      const css = typeof styles === 'function'
        ? styles({ ...theme }, props)
        : styles;
      return StyleSheet.create(css);
    }, [props, theme]);
  };
```
